ofctrl/ofSwitch.go

```diff
@@ -460,7 +462,15 @@ func (self *OFSwitch) publishMessage(xID uint32, result MessageResult) {
 		defer self.txLock.Unlock()
 		ch, found := self.txChans[xID]
 		if found {
-			ch <- result
+			select {
+			case ch <- result:
+			case <-time.After(messageTimeout):
```

Comment:
Thanks for the quick fix. I still have concern that the routine blocks for 13 seconds with a lock, which could lead to other functions cannot proceed, it's like a JVM stop-the-world and unpredictable. Could we confirm whether a buffered channel can avoid blocking here? For example, if we can confirm the outbound and inbound message is always 1 to 1 interaction, then 1 buffer is enough and it doesn't need to wait?

Reply:
There are several things to consider:
1. the channel should not be closed when it's in the map
2. the channel should have active consumer when `publishMessage` is adding items to it
3. `listenReply` (the consumer of the channel) should terminate eventually regardless the bundle is aborted or completed

So I'm thinking:
1. In `Transaction.Begin/Abort/Commit`, it should always unsubscribe first, then close the `controlReplyCh` channel, so that `publishMessage` won't send on closed channel
2. In `Transaction.listenReply`, it should always wait for the `controlReplyCh` channel to be closed first, so that the channel always has active consumer
3. `publishMessage` could be kept as is because as long as the channel is present, it's guaranteed not closed and having consumer.
4. `controlReplyCh` could have some buffer to avoid `publishMessage` holds the lock too long.
5. `controlIntCh` don't need to be closed as no consumer is waiting for that, and `listenReply` might send message to it.

Reply:
Thanks for the suggestions. I've updated the code.